shared/utility/actuation/InverseKinematics.hpp

```diff
@@ -42,177 +42,143 @@ namespace utility::actuation::kinematics {
-        Scalar length             = targetLeg.norm();
-        if (length > maxLegLength) {
-            targetLeg = targetLeg * maxLegLength / length;
-            length    = targetLeg.norm();
+        // Apply offsets to Htf and clamp to limits
+        const Eigen::Matrix<Scalar, 3, 1> ankle_x   = Htf.matrix().template leftCols<1>().template head<3>();
+        const Eigen::Matrix<Scalar, 3, 1> ankle_y   = Htf.matrix().template middleCols<1>(1).template head<3>();
+        const Eigen::Matrix<Scalar, 3, 1> ankle_pos = Htf.translation();
+
+        const Eigen::Matrix<Scalar, 3, 1> hip_offset(LENGTH_BETWEEN_LEGS * 0.5,
+                                                     HIP_OFFSET_X,
+                                                     DISTANCE_FROM_BODY_TO_HIP_JOINT);
+        Eigen::Matrix<Scalar, 3, 1> target_leg = ankle_pos - hip_offset;
+
+        const Scalar max_leg_length = UPPER_LEG_LENGTH + LOWER_LEG_LENGTH;
+        Scalar length               = target_leg.norm();
+        if (length > max_leg_length) {
+            target_leg = target_leg * max_leg_length / length;
+            length     = target_leg.norm();
         }
-        const Scalar sqrLength   = length * length;
-        const Scalar sqrUpperLeg = UPPER_LEG_LENGTH * UPPER_LEG_LENGTH;
-        const Scalar sqrLowerLeg = LOWER_LEG_LENGTH * LOWER_LEG_LENGTH;
-
-        const Scalar cosKnee = (sqrUpperLeg + sqrLowerLeg - sqrLength) / (2.0 * UPPER_LEG_LENGTH * LOWER_LEG_LENGTH);
-        knee                 = std::acos(std::max(std::min(cosKnee, Scalar(1.0)), Scalar(-1.0)));
-
-        const Scalar cosLowerLeg = (sqrLowerLeg + sqrLength - sqrUpperLeg) / (2.0 * LOWER_LEG_LENGTH * length);
-        const Scalar lowerLeg    = std::acos(std::max(std::min(cosLowerLeg, Scalar(1.0)), Scalar(-1.0)));
 
-        const Scalar phi2 = std::acos(targetLeg.dot(ankleY) / length);
-
-        anklePitch = lowerLeg + phi2 - M_PI_2;
-
-        const Eigen::Matrix<Scalar, 3, 1> unitTargetLeg = targetLeg / length;
-
-        Eigen::Matrix<Scalar, 3, 1> hipX = ankleY.cross(unitTargetLeg);
-        Scalar hipXLength                = hipX.norm();
-        if (hipXLength > 0) {
-            hipX /= hipXLength;
+        // Knee pitch
+        const Scalar sqr_length    = length * length;
+        const Scalar sqr_upper_leg = UPPER_LEG_LENGTH * UPPER_LEG_LENGTH;
+        const Scalar sqr_lower_leg = LOWER_LEG_LENGTH * LOWER_LEG_LENGTH;
+        const Scalar cos_knee =
+            (sqr_upper_leg + sqr_lower_leg - sqr_length) / (2.0 * UPPER_LEG_LENGTH * LOWER_LEG_LENGTH);
+        const Scalar knee_pitch = M_PI - std::acos(std::max(std::min(cos_knee, Scalar(1.0)), Scalar(-1.0)));
+
+        // Ankle pitch
+        const Scalar cos_lower_leg = (sqr_lower_leg + sqr_length - sqr_upper_leg) / (2.0 * LOWER_LEG_LENGTH * length);
+        const Scalar lower_leg     = std::acos(std::max(std::min(cos_lower_leg, Scalar(1.0)), Scalar(-1.0)));
```

Comment:
If you have to use this (because you can't use the dot product numerically stable version) there is a function acos_clamped in the angle utility that does all this for you